HW2/P2/parallel_vector.pyx

```diff
@@ -107,9 +140,41 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
    ##########
    with nogil:
        for r in range(repeat):
-           for idx in range(src.shape[0]):
+           for idx in prange(src.shape[0],num_threads=5):
+
+               # don't do anything if source and dest same
+               if src[idx] == dest[idx]:
+                   continue
+
+               # LOCK
+               # acquire locks in order
+               if src[idx]/N < dest[idx]/N:
+                   acquire(&locks[src[idx]/N])
+                   acquire(&locks[dest[idx]/N])
+               elif dest[idx]/N < src[idx]/N:
+                   acquire(&locks[dest[idx]/N])
+                   acquire(&locks[src[idx]/N])
+               # don't double lock
+               elif src[idx]/N == dest[idx]/N:
+                   acquire(&locks[src[idx]/N])
```

Comment:
Nice point. When I first approach the problem, I used double lock and the efficiency is definitely low.